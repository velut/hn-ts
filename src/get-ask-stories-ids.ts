import fetch from "isomorphic-unfetch";
import { hnApiBaseUrl } from "./hn-api-base-url";

/**
 * `getAskStoriesIds` returns the list of up to 200 IDs of the "Ask HN" stories.
 *
 * @see {@link https://github.com/HackerNews/API#ask-show-and-job-stories}
 */
export async function getAskStoriesIds(): Promise<number[]> {
  const endpoint = `${hnApiBaseUrl}/askstories.json`;
  const res = await fetch(endpoint);
  const ids = await res.json();
  return ids;
}
