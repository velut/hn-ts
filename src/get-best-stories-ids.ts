import fetch from "isomorphic-unfetch";
import { hnApiBaseUrl } from "./hn-api-base-url";

/**
 * `getBestStoriesIds` returns the list of up to 500 IDs of the best HN stories.
 *
 * @see {@link https://github.com/HackerNews/API#new-top-and-best-stories}
 */
export async function getBestStoriesIds(): Promise<number[]> {
  const endpoint = `${hnApiBaseUrl}/beststories.json`;
  const res = await fetch(endpoint);
  const ids = await res.json();
  return ids;
}
