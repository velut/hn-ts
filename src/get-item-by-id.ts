import { getRawItemById } from "./get-raw-item-by-id";
import { Item } from "./item";
import { normalizeRawItem } from "./normalize-raw-item";

/**
 * `getItemById` returns the `Item` with the given ID.
 * If the item is not available, `getItemById` returns `undefined`.
 *
 * @see {@link Item}
 * @see {@link https://github.com/HackerNews/API#items}
 */
export async function getItemById({
  id,
}: {
  id: number;
}): Promise<Item | undefined> {
  const rawItem = await getRawItemById({ id });
  if (!rawItem) {
    return undefined;
  }

  return normalizeRawItem({ rawItem });
}
