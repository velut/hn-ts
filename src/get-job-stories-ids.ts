import fetch from "isomorphic-unfetch";
import { hnApiBaseUrl } from "./hn-api-base-url";

/**
 * `getJobStoriesIds` returns the list of up to 200 IDs of the job HN stories.
 *
 * @see {@link https://github.com/HackerNews/API#ask-show-and-job-stories}
 */
export async function getJobStoriesIds(): Promise<number[]> {
  const endpoint = `${hnApiBaseUrl}/jobstories.json`;
  const res = await fetch(endpoint);
  const ids = await res.json();
  return ids;
}
