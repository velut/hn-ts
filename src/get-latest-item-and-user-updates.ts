import fetch from "isomorphic-unfetch";
import { hnApiBaseUrl } from "./hn-api-base-url";
import { Updated } from "./updated";

/**
 * `getLatestItemAndUserUpdates` returns the latest updates for items and users.
 *
 * @see {@link https://github.com/HackerNews/API#changed-items-and-profiles}
 */
export async function getLatestItemAndUserUpdates(): Promise<Updated> {
  const endpoint = `${hnApiBaseUrl}/updates.json`;
  const res = await fetch(endpoint);
  const updates = await res.json();
  return updates;
}
