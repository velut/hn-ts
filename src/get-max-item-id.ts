import fetch from "isomorphic-unfetch";
import { hnApiBaseUrl } from "./hn-api-base-url";

/**
 * `getMaxItemId` returns the ID of the most recent item available from the API.
 *
 * @see {@link https://github.com/HackerNews/API#max-item-id}
 */
export async function getMaxItemId(): Promise<number> {
  const endpoint = `${hnApiBaseUrl}/maxitem.json`;
  const res = await fetch(endpoint);
  const id = await res.json();
  return id;
}
