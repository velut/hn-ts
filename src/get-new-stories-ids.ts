import fetch from "isomorphic-unfetch";
import { hnApiBaseUrl } from "./hn-api-base-url";

/**
 * `getNewStoriesIds` returns the list of up to 500 IDs of the newer HN stories.
 *
 * @see {@link https://github.com/HackerNews/API#new-top-and-best-stories}
 */
export async function getNewStoriesIds(): Promise<number[]> {
  const endpoint = `${hnApiBaseUrl}/newstories.json`;
  const res = await fetch(endpoint);
  const ids = await res.json();
  return ids;
}
