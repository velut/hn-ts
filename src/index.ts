/**
 * This package TODO:
 *
 * @packageDocumentation
 */

export * from "./get-ask-stories-ids";
export * from "./get-best-stories-ids";
export * from "./get-item-by-id";
export * from "./get-job-stories-ids";
export * from "./get-max-item-id";
export * from "./get-new-stories-ids";
export * from "./get-raw-item-by-id";
export * from "./get-raw-user-by-id";
export * from "./get-show-stories-ids";
export * from "./get-top-stories-ids";
export * from "./get-user-by-id";
export * from "./hn-api-base-url";
export * from "./item";
export * from "./item-type";
export * from "./raw-item";
export * from "./raw-item-type";
export * from "./raw-user";
export * from "./user";
