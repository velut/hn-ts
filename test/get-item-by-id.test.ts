/** @jest-environment setup-polly-jest/jest-environment-node */
// See https://netflix.github.io/pollyjs/#/test-frameworks/jest-jasmine?id=supported-test-runners

import NodeHttpAdapter from "@pollyjs/adapter-node-http";
import FSPersister from "@pollyjs/persister-fs";
import * as path from "path";
import { setupPolly } from "setup-polly-jest";
import { getItemById } from "../src";

describe("getItemById", () => {
  setupPolly({
    adapters: [NodeHttpAdapter],
    persister: FSPersister,
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, "./__recordings__"),
      },
    },
    recordFailedRequests: true,
  });

  it("returns an existing story item", async () => {
    expect.assertions(5);

    const item = await getItemById({ id: 27107832 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(27107832);
    expect(item!.author).toEqual("velut");
    expect(item!.type).toEqual("story");
    expect(item).toMatchSnapshot();
  });

  it("returns an existing comment item", async () => {
    expect.assertions(5);

    const item = await getItemById({ id: 27121843 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(27121843);
    expect(item!.author).toEqual("velut");
    expect(item!.type).toEqual("comment");
    expect(item).toMatchSnapshot();
  });

  it("returns an existing 'ask hn' story item", async () => {
    expect.assertions(5);

    const item = await getItemById({ id: 121003 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(121003);
    expect(item!.type).toEqual("story");
    expect(item!.title!.startsWith("Ask HN")).toEqual(true);
    expect(item).toMatchSnapshot();
  });

  it("returns an existing 'show hn' story item", async () => {
    expect.assertions(5);

    const item = await getItemById({ id: 25496976 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(25496976);
    expect(item!.type).toEqual("story");
    expect(item!.title!.startsWith("Show HN")).toEqual(true);
    expect(item).toMatchSnapshot();
  });

  it("returns an existing job item", async () => {
    expect.assertions(4);

    const item = await getItemById({ id: 192327 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(192327);
    expect(item!.type).toEqual("job");
    expect(item).toMatchSnapshot();
  });

  it("returns an existing poll item", async () => {
    expect.assertions(4);

    const item = await getItemById({ id: 126809 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(126809);
    expect(item!.type).toEqual("poll");
    expect(item).toMatchSnapshot();
  });

  it("returns an existing poll option item", async () => {
    expect.assertions(4);

    const item = await getItemById({ id: 160705 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(160705);
    expect(item!.type).toEqual("pollopt");
    expect(item).toMatchSnapshot();
  });

  it("returns a deleted item", async () => {
    expect.assertions(4);

    const item = await getItemById({ id: 27121851 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(27121851);
    expect(item!.deleted).toEqual(true);
    expect(item).toMatchSnapshot();
  });

  it("returns a dead item", async () => {
    expect.assertions(4);

    const item = await getItemById({ id: 29013059 });

    expect(item).toBeDefined();
    expect(item!.id).toEqual(29013059);
    expect(item!.dead).toEqual(true);
    expect(item).toMatchSnapshot();
  });

  it("returns undefined if an item with the given ID is not found", async () => {
    expect.assertions(1);

    const item = await getItemById({ id: 0 });

    expect(item).toBeUndefined();
  });
});
