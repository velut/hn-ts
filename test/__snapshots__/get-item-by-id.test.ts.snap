// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getItemById returns a dead item 1`] = `
Object {
  "author": "clement189",
  "childrenIds": undefined,
  "dead": true,
  "deleted": false,
  "id": 29013059,
  "numChildren": undefined,
  "parentId": undefined,
  "pollId": undefined,
  "pollOptionsIds": undefined,
  "score": 1,
  "text": undefined,
  "timestamp": "2021-10-27T13:31:37.000Z",
  "title": "Move from one page to another without loading",
  "type": "story",
  "url": "https://dev.to/clementgaudiniere/move-from-one-page-to-another-without-loading-jlb",
}
`;

exports[`getItemById returns a deleted item 1`] = `
Object {
  "author": undefined,
  "childrenIds": undefined,
  "dead": false,
  "deleted": true,
  "id": 27121851,
  "numChildren": undefined,
  "parentId": 27120206,
  "pollId": undefined,
  "pollOptionsIds": undefined,
  "score": undefined,
  "text": undefined,
  "timestamp": "2021-05-11T18:41:22.000Z",
  "title": undefined,
  "type": "comment",
  "url": undefined,
}
`;

exports[`getItemById returns an existing comment item 1`] = `
Object {
  "author": "velut",
  "childrenIds": undefined,
  "dead": false,
  "deleted": false,
  "id": 27121843,
  "numChildren": undefined,
  "parentId": 27121265,
  "pollId": undefined,
  "pollOptionsIds": undefined,
  "score": undefined,
  "text": "Thank you!

Yes, I use prism-react-renderer for syntax highlighting.

I did not know about Shiki, but I'll look into it.",
  "timestamp": "2021-05-11T18:40:46.000Z",
  "title": undefined,
  "type": "comment",
  "url": undefined,
}
`;

exports[`getItemById returns an existing story item 1`] = `
Object {
  "author": "velut",
  "childrenIds": Array [
    27107840,
    27112027,
    27112015,
    27121265,
  ],
  "dead": false,
  "deleted": false,
  "id": 27107832,
  "numChildren": 7,
  "parentId": undefined,
  "pollId": undefined,
  "pollOptionsIds": undefined,
  "score": 14,
  "text": undefined,
  "timestamp": "2021-05-10T16:30:22.000Z",
  "title": "Show HN: Automatic Documentation for TypeScript Packages",
  "type": "story",
  "url": "https://www.jsdocs.io",
}
`;

exports[`getItemById returns an item with only an ID 1`] = `
Object {
  "author": undefined,
  "childrenIds": undefined,
  "dead": false,
  "deleted": false,
  "id": 0,
  "numChildren": undefined,
  "parentId": undefined,
  "pollId": undefined,
  "pollOptionsIds": undefined,
  "score": undefined,
  "text": undefined,
  "timestamp": undefined,
  "title": undefined,
  "type": "unknown",
  "url": undefined,
}
`;
